src/core/reference/src/runtime/opt_kernel/reshape.cpp

```diff
@@ -73,26 +96,39 @@ void reshape_in3(const char* in,
                  const AxisVector& in_axis_order,
                  const Shape& out_shape,
                  size_t elem_size) {
-    size_t size[3];
-    size_t in_index[3];
-    size_t* map_index[3];
-    for (size_t i = 0; i < 3; i++) {
-        size[i] = in_shape[in_axis_order[i]];
-        map_index[in_axis_order[i]] = &in_index[i];
-    }
-    for (in_index[0] = 0; in_index[0] < size[0]; ++in_index[0]) {
-        for (in_index[1] = 0; in_index[1] < size[1]; ++in_index[1]) {
-            for (in_index[2] = 0; in_index[2] < size[2]; ++in_index[2]) {
-                // clang-format off
-                    memcpy(out,
-                           in + (*map_index[0] * in_shape[1] * in_shape[2] +
-                                 *map_index[1] * in_shape[2] +
-                                 *map_index[2]) * elem_size,
-                           elem_size);
+    size_t num_elements = shape_size(in_shape);
+    if (num_elements <= get_threshold()) {
+        size_t i, j, k;
+        size_t* in_indexes[3];
+        in_indexes[in_axis_order[0]] = &i;
+        in_indexes[in_axis_order[1]] = &j;
+        in_indexes[in_axis_order[2]] = &k;
+
+        for (i = 0; i < out_shape[0]; i++) {
```

Comment:
Can you explain why  pointers to indexes are used?
The alternative solution could be like in each loop the sub-offset could be calculated per iteration like:
```suggestion
        for (i = 0; i < out_shape[0]; i++) {
            size_t sub_offset_i = i * in_shape[1];
```

or 

```suggestion
        size_t sub_offset_i = 0;
        for (i = 0; i < out_shape[0]; i++) {
           // add in_shape every iteration to avoid multiplication and it should be eq to i * in_shape[1]
            sub_offset_i += in_shape[1];
```

Then in `copy_element` just sum all of them:
```cpp
copy_element(out, in +(sub_offset_i + sub_offset_j + ..., elem_size);
```
I think working fundamental types directly should be faster then using access to table and then via pointer to value  (indirect access).

But maybe current solution better(faster)?